Provide support for LZCNT and TZCNT in X86Ops.ins

VidushiGit opened the pull request.

0xdaryl commented:
@BradleyWood : please review

0xdaryl commented:
Thanks for the contribution.  Could you add encodings for the other register sizes as well please?  i.e., 16, 32, and 64-bit

BradleyWood commented:
Can you add [binary encoding tests](https://github.com/eclipse/omr/blob/de6212d59d65524f189b3438b1c78ba74a524358/fvtest/compilerunittest/x/BinaryEncoder.cpp#L153) like the following?

`
INSTANTIATE_TEST_CASE_P(GeneralPurposeRegRegTest, XRegRegEncEncodingTest, ::testing::ValuesIn(*TRTest::MakeVector<std::tuple<TR::InstOpCode::Mnemonic, TR::RealRegister::RegNum, TR::RealRegister::RegNum, OMR::X86::Encoding, TRTest::BinaryInstruction>>(
    std::make_tuple(TR::InstOpCode::LZCNTRegReg,   TR::RealRegister::eax, TR::RealRegister::ecx, OMR::X86::Default, "<binary encoding>"),
    std::make_tuple(TR::InstOpCode::TZCNTRegReg,   TR::RealRegister::ecx, TR::RealRegister::eax, OMR::X86::Default, "<binary encoding>")
)));
`

VidushiGit commented:
> Thanks for the contribution. Could you add encodings for the other register sizes as well please? i.e., 16, 32, and 64-bit

Yes, I will add support for the 16 and 32- bit also. Current implementation is only supporting 64 bits. Thanks

VidushiGit commented:
Hi Bradley,
I have added the support for 16 and 32-bit registers for the instructions.

Also I have added the test cases for the instructions as suggested. Currently I have added EAX,ECX register pair(in the same order) for testing. Please let us know if we need some other test cases added.

BradleyWood commented on fvtest/compilerunittest/x/BinaryEncoder.cpp:
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
 
+INSTANTIATE_TEST_CASE_P(GeneralPurposeRegRegTest, XRegRegEncEncodingTest, ::testing::ValuesIn(*TRTest::MakeVector<std::tuple<TR::InstOpCode::Mnemonic, TR::RealRegister::RegNum, TR::RealRegister::RegNum, OMR::X86::Encoding, TRTest::BinaryInstruction>>(
+    std::make_tuple(TR::InstOpCode::LZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbdc1"),
+    std::make_tuple(TR::InstOpCode::LZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f34c0fbdc1"),
```
`f34c0fbdc1` equates to `lzcnt  r8,rcx` which does not match the specified registers

BradleyWood commented on fvtest/compilerunittest/x/BinaryEncoder.cpp:
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
+    std::make_tuple(TR::InstOpCode::LZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f34c0fbdc1"),
+
+    std::make_tuple(TR::InstOpCode::TZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbcc1"),
+    std::make_tuple(TR::InstOpCode::TZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f34c0fbcc1")
```
'f34c0fbcc1' equates to `tzcnt  r8,rcx` which does not match the specified registers

VidushiGit commented:
Corrected the test cases' binary encoding string

0xdaryl commented on compiler/x/codegen/X86Ops.ins:
```diff
@@ -5144,6 +5159,21 @@ INSTRUCTION(TEST8MemReg, test,
             PROPERTY1(IA32OpProp1_LongSource | IA32OpProp1_LongTarget | IA32OpProp1_FusableCompare),
             FEATURES(0)),
+INSTRUCTION(TZCNT2RegReg, tzcnt,
+            BINARY(VEX_L___, VEX_vNONE, PREFIX_F3, REX__, ESCAPE_0F__, 0xbc, 0, ModRM_RM__, Immediate_0),
```
Same comment as above for the operand size prefix.

0xdaryl commented on fvtest/compilerunittest/x/BinaryEncoder.cpp:
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
+    std::make_tuple(TR::InstOpCode::LZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f3480fbdc1"),
+
+    std::make_tuple(TR::InstOpCode::TZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbcc1"),
+    std::make_tuple(TR::InstOpCode::TZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f3480fbcc1")
```
Can you add tests for the 16-bit versions as well?

VidushiGit replied:
Added tests for 16-bit version of the instruction

BradleyWood commented on compiler/x/codegen/X86Ops.ins:
```diff
@@ -2545,6 +2545,21 @@ INSTRUCTION(LEA8RegMem, lea,
             PROPERTY1(IA32OpProp1_LongSource | IA32OpProp1_LongTarget | IA32OpProp1_SourceIsMemRef),
             FEATURES(0)),
+INSTRUCTION(LZCNT2RegReg, lzcnt,
+            BINARY(VEX_L___, VEX_vNONE, PREFIX_F3, REX__, ESCAPE_0F__, 0xbd, 0, ModRM_RM__, Immediate_0),
```
I had a look over the code, as it currently stands you cannot use both prefixes.

0xdaryl replied:
PR #7020 was merged which provides the capability of specifying multiple prefixes.  @VidushiGit, you should pick up that change and use it here.

VidushiGit replied:
@0xdaryl Pulled the latest OMR code changes and added the support/usage for multiple instruction prefixes

BradleyWood commented:
jenkins build all

BradleyWood commented:
MacOS failure is #6516
